File: Swift_2016/23-block的回顾/23-block的回顾/ViewController.m
//
//  ViewController.m
//  23-block的回顾
//
//  Created by ld on 2023/2/21.
//

#import "ViewController.h"
#import "HttpTool.h"
@interface ViewController ()

@property (nonatomic,strong) HttpTool *tools;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.tools = [[HttpTool alloc] init];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    
    /*
     __weak修饰的弱引用,如果指向的对象销毁,那么指针会立马指向nil(0x0)
     __unsafe_unretained修饰的弱引用,如果指向的对象销毁,那么指针依然指向之前的内存地址,很容易产生'野指针'/'僵尸对象'
     */
    
    __weak ViewController *weakSelf = self;
    [self.tools loadData:^(NSString * _Nonnull jsonData) {
         NSLog(@"在ViewController拿到数据:%@", jsonData);
        weakSelf.view.backgroundColor = [UIColor redColor];
    }];
}

- (void)dealloc {
    NSLog(@"ViewController -- dealloc");
}

@end

File: Swift_2016/24-闭包的使用/24-闭包的使用/ViewController.swift
//
//  ViewController.swift
//  24-闭包的使用
//
//  Created by ld on 2023/2/21.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }
    
    var tools : HttpTool = HttpTool()
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        tools.loadData { jsonData in
             print("在ViewController拿到数据:\(jsonData)")
            self.view.backgroundColor = UIColor.red
        }
    }

}

